### ISSUE TITLE
definedIn filter false-positives on PHP builtins called from namespaced code

### ISSUE THREAD

**Issue by aldemeery** [spaze/phpstan-disallowed-calls#386]
## Context

This is a follow-up to the fix for the `$isBuiltIn` check added to `definedInMatches()`. That fix correctly handles the case where `hasFunction($name, $scope)` succeeds and `isBuiltin()` returns `true`. However, there is a second code path that bypasses it.

## Problem

When a PHP builtin (e.g. `sprintf`) is called from namespaced code (e.g. `namespace Tests\Architecture`), PHP-Parser / PHPStan sets `namespacedName` on the AST node to a `Name\FullyQualified(['Tests', 'Architecture', 'sprintf'])`. `TypeResolver::getNames()` returns this as the first candidate name.

`DisallowedFunctionRuleErrors::getErrors()` then calls `hasFunction(Tests\Architecture\sprintf, $scope)`. This fails — no such namespaced function exists — and the `else` branch runs:

```php
} else {
    $definedIn = null;
    $isBuiltIn = false;  // ← never corrected
}
```

`$isBuiltIn` stays `false`, so `definedInMatches()`'s `elseif ($isBuiltIn)` branch is never reached. `$definedIn === null` then triggers the conservative `return true`, and the builtin is incorrectly flagged.

The loop in `get()` returns early after the first name produces errors, so the plain `sprintf` name (which would resolve correctly) is never tried.

## Possible Fix

In the `else` branch of `getErrors()`, when `hasFunction` fails on a non-unqualified name, check whether the last segment resolves to a global builtin:

```php
// Spaze\PHPStan\Rules\Disallowed\RuleErrors\DisallowedFunctionRuleErrors:getErrors()

} else {
    $definedIn = null;
    $isBuiltIn = false;

    // If hasFunction failed on a namespace-qualified name, check whether the
    // base name is a global builtin (e.g. Tests\Foo\sprintf → sprintf).
    if (!$name->isUnqualified()) {
        $global = new Name($name->getLast());
        if ($this->reflectionProvider->hasFunction($global, null)) {
            $isBuiltIn = $this->reflectionProvider->getFunction($global, null)->isBuiltin();
        }
    }
}
```

Note: `isQualified()` is insufficient here — `Name\FullyQualified` returns `false` for `isQualified()` even with multiple parts. `!isUnqualified()` correctly covers plain qualified, fully qualified, and relative name types.

## Why it's safe

- Only runs in the `else` branch (when `hasFunction` already failed for the namespaced form)
- `isBuiltIn` is only set to `true` if the base name resolves to an actual builtin — PHP disallows redefining builtin names globally, so `isBuiltin()` is always `true` for real builtins and `false` for user-land functions
- If the base name has no global function at all, `$isBuiltIn` remains `false` and conservative behaviour is fully preserved

**Comment by spaze** [comment]
Thank you for the follow-up issue, and sorry I didn't get it right the first time 😅

The problem seems to be a bit deeper than with the built-in functions:

```php
namespace {
	function func()
	{
	}
}

namespace Foo {
	func();
}
```

`func()` will also be flagged (as `Foo\func()`). I'll check this, I think I have already seen something similar.

**Comment by spaze** [comment]
I *think* I have finally fixed the problem: if the reflection provider doesn't have the function, just skip it, that way the "wrongly namespaced function" will be skipped - that's the same approach as for methods for example. I'm quite confident this fixes it, but looking back 😅, can I ask you to test the code before I merge and release it? If you don't want to do any composer magic, you can replace just one file, the rest is new or updated tests.

**Comment by aldemeery** [comment]
@spaze First of all, I can't thank you enough for how timely your response has been, and I am genuinely impressed by how dedicated you are!

> can I ask you to test the code before I merge and release it? If you don't want to do any composer magic, you can replace just one file, the rest is new or updated tests.

Of course! I will test this ASAP!

**Comment by aldemeery** [comment]
Confirmed! 👍🏻  Works perfectly for all our use cases. Thanks again!

**Comment by spaze** [comment]
Thank you very much @aldemeery for your nice words, means a lot to me! :-) Also thanks for testing the fix, it's merged now, and released in [4.9.2](https://github.com/spaze/phpstan-disallowed-calls/releases/tag/v4.9.2).
